### PR TITLE
syntax: Apply `x as usize < y` recovery to type ascription as well

### PR DESCRIPTION
**Pull request by petrochenkov**
Also correct spans, add some comments.

r? @estebank

### CODE REVIEW

File: src/test/ui/issue-22644.stderr [hunk 1]
```diff
@@ -1,24 +1,66 @@
...
 
-error: aborting due to 2 previous errors
+error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
+  --> $DIR/issue-22644.rs:20:31
+   |
+20 |     println!("{}{}", a: usize < long_name, long_name);
+   |                               ^---------------------
+   |                               |
+   |                               interpreted as generic arguments
+   |                               not interpreted as comparison
```
Comment (estebank) [inline src/test/ui/issue-22644.stderr]: Will look at the code this afternoon to see what change would be needed for this suggestion to be implemented, but I feel the output should be:

```
   |
20 |     println!("{}{}", a: usize < long_name, long_name);
   |                               ^ -------------------- interpreted as generic arguments
   |                               |
   |                               not interpreted as comparison
```

**Comment by alexcrichton** [comment]
Assigning @nrc just to make sure someone's assigned here, but looks like @estebank has review taken care of

**Comment by petrochenkov** [comment]
@estebank 
Updated with the span fix.

**Comment by petrochenkov** [comment]
Rebased.
@bors r=estebank

**Pull request merged**